src/main/java/school/lemon/changerequest/java/collections/hw1/ExtendedListImpl.java

```diff
@@ -0,0 +1,709 @@
+                    if (ExtendedListImpl.this.container[begin + i] == null)
+                        return i;
+            } else {
+                for (int i = 0; i < size; i++)
+                    if (o.equals(ExtendedListImpl.this.container[begin + i]))
+                        return i;
+            }
+            return -1;
+        }
+
+        @Override
+        public int lastIndexOf(Object o) {
+            checkConcurrentModification();
+            if (o == null) {
+                for (int i = size - 1; i >= 0; i--)
+                    if (ExtendedListImpl.this.container[begin + i] == null)
+                        return i;
+            } else {
+                for (int i = size - 1; i >= 0; i--)
+                    if (o.equals(ExtendedListImpl.this.container[begin + i]))
+                        return i;
+            }
+            return -1;
+        }
+
+        @Override
+        public ListIterator<E> listIterator() {
+            checkConcurrentModification();
+            return new ListItr(this);
+        }
+
+        @Override
+        public ListIterator<E> listIterator(int index) {
+            checkConcurrentModification();
+            isCorrectIndexForAdd(index);
+            return new ListItr(this, index);
+        }
+
+        @Override
+        public List<E> subList(int fromIndex, int toIndex) {
+            checkSubListIndexes(fromIndex, toIndex);
+            checkConcurrentModification();
+            return new SubList(this, begin, fromIndex, toIndex);
+        }
+
+        @Override
+        public String toString() {
+            StringBuilder result = new StringBuilder("[");
+            if (size > 0)
+                result.append(ExtendedListImpl.this.container[begin]);
+            for (int i = 1; i < size; i++)
+                result.append(", ").append(ExtendedListImpl.this.container[begin + i]);
+            result.append("]");
+            return result.toString();
+        }
+
+        private void isCorrectIndex(int index) {
+            if (index < 0 || index >= size)
+                throw new IndexOutOfBoundsException("Incorrect index:" + index);
+        }
+
+        private void isCorrectIndexForAdd(int index) {
+            if (index < 0 || index > size)
+                throw new IndexOutOfBoundsException("Incorrect index:" + index);
+        }
+
+        private void checkSubListIndexes(int fromIndex, int toIndex) {
+            if (fromIndex < 0 || toIndex > size || fromIndex > toIndex)
+                throw new IndexOutOfBoundsException("Incorrect index:");
+        }
+
+        private void checkConcurrentModification() {
+            if (modificationCoef != ExtendedListImpl.this.modificationCoef)
+                throw new ConcurrentModificationException();
+        }
+
+    }
+
+
+    /**********************************************************************************/
+
+    private class ListItr extends Itr implements ListIterator<E> {
```

Comment:
Same for name here. `ListIteratorImpl` or `BidirectionalIterator` would be better here.